File: src/main/Java/boj_step15/boj17103.java
import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

public class boj17103 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(reader.readLine());
        boolean[] isPrime = new boolean[1_000_000];
        Arrays.fill(isPrime, true);
        isPrime[0] = false; isPrime[1] = false;
        List<Integer> primes = new ArrayList<>();
        for (int i = 2; i < isPrime.length; i++) {
            if (!isPrime[i]) continue;
            primes.add(i);
            for (int j = i * 2; j < isPrime.length; j += i) isPrime[j] = false;
        }
        while (n-- > 0) {
            int partition = Integer.parseInt(reader.readLine());
            AtomicInteger count = new AtomicInteger(0);
            primes.forEach(it -> {
                if (it > partition / 2) return;
                if (isPrime[partition - it]) count.getAndIncrement();
            });
            writer.append(count + "\n");
        }
        writer.close();
    }
}

File: src/main/boj_step15/boj17103.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val n = r.readLine().toInt()
    val isPrimes = BooleanArray(1_000_000) { true }
    isPrimes[0] = false
    isPrimes[1] = false
    for (i in 2 until isPrimes.size) {
        if (isPrimes[i] == false) continue
        for (j in i * 2 until isPrimes.size step (i)) isPrimes[j] = false
    }
    val primes = mutableListOf<Int>()
    isPrimes.forEachIndexed { index, it -> if (it) primes.add(index) }
    repeat(n) {
        val partition = r.readLine().toInt()
        var count = 0
        primes.forEach {
            if (it > partition / 2) return@forEach
            if (isPrimes[partition - it]) count++
        }
        w.appendLine(count.toString())
    }
    w.close()
}




